File: ganesha-rs/ganesha4/merge_sort.py
def merge_sort(arr):
    """
    Sorts an array using the Merge Sort algorithm.

    Parameters:
        arr (list): The list of comparable elements to sort.

    Returns:
        list: A new sorted list.
    """
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    return _merge(left_half, right_half)

def _merge(left, right):
    """Helper function to merge two sorted lists."""
    merged = []
    i = j = 0

    # Merge the two halves
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    # Append any remaining elements
    merged.extend(left[i:])
    merged.extend(right[j:])

    return merged

if __name__ == "__main__":
    import random
    sample = [random.randint(0, 100) for _ in range(20)]
    print("Unsorted:", sample)
    sorted_list = merge_sort(sample)
    print("Sorted:  ", sorted_list)

File: ganesha-rs/ganesha4/merge_sort.rs
fn merge_sort<T: Ord + Copy>(arr: &mut [T]) {
    if arr.len() <= 1 {
        return;
    }

    let mid = arr.len() / 2;
    let mut left = arr[..mid].to_vec();
    let mut right = arr[mid..].to_vec();

    merge_sort(&mut left);
    merge_sort(&mut right);

    merge(arr, &left, &right);
}

fn merge<T: Ord + Copy>(arr: &mut [T], left: &[T], right: &[T]) {
    let mut i = 0;
    let mut j = 0;
    let mut k = 0;

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            arr[k] = left[i];
            i += 1;
        } else {
            arr[k] = right[j];
            j += 1;
        }
        k += 1;
    }

    while i < left.len() {
        arr[k] = left[i];
        i += 1;
        k += 1;
    }

    while j < right.len() {
        arr[k] = right[j];
        j += 1;
        k += 1;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merge_sort() {
        let mut arr = [5, 2, 8, 1, 9, 4, 7, 3, 6];
        merge_sort(&mut arr);
        assert_eq!(arr, [1, 2, 3, 4, 5, 6, 7, 8, 9]);
    }

    #[test]
    fn test_merge_sort_empty() {
        let mut arr: [i32; 0] = [];
        merge_sort(&mut arr);
        assert_eq!(arr, []);
    }

    #[test]
    fn test_merge_sort_single() {
        let mut arr = [5];
        merge_sort(&mut arr);
        assert_eq!(arr, [5]);
    }
}

fn main() {
    let mut arr = [5, 2, 8, 1, 9, 4, 7, 3, 6];
    println!("Unsorted array: {:?}", arr);
    merge_sort(&mut arr);
    println!("Sorted array: {:?}", arr);
}
